### PR TITLE
[metrics] Generate the dashboard json programmatically

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Generate the grafana configs programmatically to allow easier editing. Changes rolled in from https://github.com/ray-project/ray/pull/29910

### CODE REVIEW

File: dashboard/modules/metrics/grafana_dashboard_factory.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+
+@dataclass
+class Target:
+    expr: str
+    legend: str
+
+
+@dataclass
+class Panel:
```
Comment (rkooo567) [inline dashboard/modules/metrics/grafana_dashboard_factory.py:18]: Can you write a simple docstring for Panel and Target? Also it'd be nice to mention the title here should match the file inside Metrics.tsx

File: dashboard/modules/metrics/metrics_head.py [hunk 2]
```diff
@@ -249,19 +255,14 @@ def _create_default_grafana_configs(self):
...
-                        os.path.join(self._grafana_dashboard_output_dir, file),
-                    )
+        with open(
+            os.path.join(
+                dashboards_path,
+                "default_grafana_dashboard.json",
+            ),
+            "w",
+        ) as f:
+            f.write(generate_grafana_dashboard())
```
Comment (rkooo567) [inline dashboard/modules/metrics/metrics_head.py:265]: Previously, we write all files inside `grafana_dashboards_dir ` here. Should we still keep the behavior? cc @alanwguo

File: dashboard/modules/metrics/grafana_dashboard_factory.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            }
+        )
+        template["gridPos"]["y"] = i // 2
+        template["gridPos"]["x"] = 12 * (i % 2)
+        template["yaxes"][0]["format"] = panel.unit
+        panels.append(template)
+    return panels
+
+
+def _generate_targets(panel):
```
Comment (rkooo567) [inline dashboard/modules/metrics/grafana_dashboard_factory.py:404]:
```suggestion
def _generate_targets(panel: Panel):
```

**Review by rkooo567 — COMMENTED** [review]
Is there any good way to test this btw? (Should we compare the output of `generate_grafana_dashboard` to other pre-created config?)

File: dashboard/modules/metrics/metrics_head.py (hunk 2 above)
Comment (ericl) [reply dashboard/modules/metrics/metrics_head.py:265]: I'd prefer we generate all dashboards programmatically, which is why I removed this.

**Comment by ericl** [comment]
> Is there any good way to test this btw? (Should we compare the output of generate_grafana_dashboard to other pre-created config?)

I did this manually for now. We may want to materialize the output for auditing purposes in the future, but I think for simple templating we're ok as is.

**Review by alanwguo — APPROVED** [review]
Nice. I can add global filters as a follow-up, when I merge in #29833

File: dashboard/modules/metrics/metrics_head.py [hunk 4]
```diff
@@ -249,19 +255,14 @@ def _create_default_grafana_configs(self):
...
-                for file in files:
-                    shutil.copy2(
-                        os.path.join(root, file),
-                        os.path.join(self._grafana_dashboard_output_dir, file),
-                    )
+        with open(
+            os.path.join(
+                dashboards_path,
+                "default_grafana_dashboard.json",
+            ),
```
Comment (alanwguo) [inline dashboard/modules/metrics/metrics_head.py:262]: This is the issue, the output path should be `_grafana_dashboard_output_dir`. That value can be overwritten via ENV VAR.

**Comment by alanwguo** [comment]
THis change should have failed `test_metrics_head.py` It fails for me locally. I wonder why it wasn't run. Any ideas?

**Pull request merged**